processors/locations.py

```diff
@@ -0,0 +1,169 @@
+
+
+def article_begin(state):
+    """ Called at the beginning of article processing """
+
+    # Delete all existing locations for this article
+    session = state["session"]  # Database session
+    url = state["url"]  # URL of the article being processed
+    session.execute(Location.table().delete().where(Location.article_url == url))
+
+    # Set of all unique locations found in article
+    state["locations"] = set()
+
+
+def article_end(state):
+    """ Called at the end of article processing """
+
+    url = state["url"]
+    session = state["session"]
+
+    locs = state.get("locations")
+    if not locs:
+        return
+
+    print(locs)
+
+    # Find all placenames mentioned in article
+    # We can use them to disambiguate addresses and street names
+    placenames = [p.name for p in locs if p.kind == "placename"]
+
+    # Get info about each location and save to database
+    for loctuple in locs:
+        loc = loctuple._asdict()
+        kind = loc["kind"]
+        coords = None
+
+        # Heimilisfang
+        if kind == "address":
+            loc["country"] = ICELAND_ISOCODE
+            info = icelandic_addr_info(loc["name"], placename_hints=placenames)
+            if info and info.get("lat_wgs84") and info.get("long_wgs84"):
+                coords = (info["lat_wgs84"], info["long_wgs84"])
+            loc["data"] = info
+
+        # Land
+        elif kind == "country":
+            code = isocode_for_country_name(loc["name"])
+            if code:
+                loc["country"] = code
+                coords = coords_for_country(code)
+
+        # Götuheiti
+        elif kind == "street":
+            # All the street names in BÍN are Icelandic
+            loc["country"] = ICELAND_ISOCODE
+            coords = coords_for_street_name(loc["name"], placename_hints=placenames)
+
+        # Örnefni
+        elif kind == "placename":
+            if loc["name"] in PLACENAME_BLACKLIST:
+                continue
+
+        if coords:
+            (loc["latitude"], loc["longitude"]) = coords
+
+        loc["article_url"] = url
+        loc["timestamp"] = datetime.utcnow()
+
+        l = Location(**loc)
+        session.add(l)
+
+
+def sentence(state, result):
+    """ Called at the end of sentence processing """
+    pass
+
+
+def Heimilisfang(node, params, result):
+    """ NP-ADDR """
+
+    children = list(node.children())
+
+    # Make sure address is in nominative case
+    def nom(w):
+        bindb = result._state["bin_db"]
+        n = bindb.lookup_nominative(w)
+        return n[0][0] if n else w
+
+    addr_nom = " ".join([nom(c.contained_text()) for c in children])
+
+    l = Loc(name=addr_nom, kind="address")
+    result._state["locations"].add(l)
+
+
+def _process(node, params, result):
+    """ Look up meaning in BÍN, add as location if in right category """
+    state = result._state
+
+    loc_fl = ["göt", "lönd", "örn"]
+    binfl2kind = dict(zip(loc_fl, ["street", "country", "placename"]))
```

Comment:
Move this to constants at the top of the file or just before `def _process(...):`